### ISSUE TITLE
Lazygit can't launch editor when adding a tag, even though it launches the editor for commits.

### ISSUE THREAD

**Issue by jbrains** [jesseduffield/lazygit#3423]
**Describe the bug**

When I use the command `T` to add a tag, lazygit doesn't launch my editor, even though it launches the editor correctly for commit messages, both for new commits and rewording existing commits.

**To Reproduce**

1. Select a commit.
1. `T`
1. Enter a tag name.

```
Fatal error: stdout is not a tty.
error: There was a problem with the editor (path to correct editor executable).
Please supply the message using either -m or -F option.
```

**Expected behavior**

Lazygit opens my editor so that I can enter a message.

**Version info:**
I ran this on 2024-03-24 at 11:29 ADT.

```
go install github.com/jesseduffield/lazygit@latest

go: downloading github.com/jesseduffield/lazygit v0.41.0
go: downloading github.com/jesseduffield/gocui v0.3.1-0.20240309085756-86e0d5a312de
go: downloading github.com/adrg/xdg v0.4.0
go: downloading github.com/gdamore/tcell/v2 v2.7.4
```

The problem persists.

**Additional context**
I haven't used tags in a long time, so I can't remember whether this has ever worked or always failed. When I run exactly the same command from my shell, everything works fine. I used `lazygit --debug` to copy/paste the exact command that lazygit runs.

**Comment by stefanhaller** [comment]
I'm confused. Lazygit is not supposed to open an editor when creating a tag. It opens its own message panel which lets you enter the tag name, and optionally a tag description; if you enter a description, it creates an annotated tag, otherwise a lightweight tag. No editor involved.

> I used `lazygit --debug` to copy/paste the exact command that lazygit runs.

And what was it? I don't see it in your bug report. You should also see the exact command in lazygit's command log panel (in the lower right of the window, you can open it with `@` if you don't see it).

**Comment by jbrains** [comment]
The command was

```
git tag -- <sha1>
```

When I run this from the command line, it opens `$EDITOR` so that I can annotate the tag.

I presume that if I don't enter a "Tag description" in `lazygit`'s window, then `git` tries to open the editor. I infer from your comment that that isn't supported.

At best, then, this is a confusing user experience. :)

Indeed, when I enter text into the "Tag description" window, `lazygit` creates an annotated tag without any problems.

It seems, then, that I can't create a lightweight tag, but only an annotated one. Why might that be?

**Comment by stefanhaller** [comment]
> The command was
> 
> ```
> git tag -- <sha1>
> ```

I find this hard to believe. No tag name? Lazygit uses the command `git tag -- <tagname> <sha1>` to create a lightweight tag.

> When I run this from the command line, it opens `$EDITOR` so that I can annotate the tag.

Again I can only assume that this wasn't the exact command line that you typed, as this would have created a lightweight tag from HEAD with the name `<sha1>`.

But what's puzzling is that you get the editor. This sounds as if you had typed `git tag -a`, but you didn't (and neither did lazygit, of course). It sounds like there might be a git config that makes `-a` the default for every tag, but I couldn't find any such option. Weird.

**Comment by jbrains** [comment]
> The command was
>
> ```
> git tag -- <sha1>
> ```
>
> I find this hard to believe. No tag name? Lazygit uses the command git tag -- <tagname> <sha1> to create a lightweight tag.

Indeed, I got that wrong. I just tried again and here is the relevant log output.

```
Mar 27 10:27:35 |DEBU| RunCommand command="git tag -- draft-0 10d8ec03c6b1bd4ffec012cd44c4df5dacd44203"
Mar 27 10:27:35 |ERRO| Fatal error: stdout is not a tty                                                                                           error: There was a problem with the editor '/home/jbrains/.local/bin/kak'.  
Please supply the message using either -m or -F option.                                                                                                                                                                    
Mar 27 10:27:35 |INFO| git tag -- draft-0 10d8ec03c6b1bd4ffec012cd44c4df5dacd44203 (98.715059ms) 
Mar 27 10:27:35 |INFO| refreshing all scopes in async mode       
```

> But what's puzzling is that you get the editor. This sounds as if you had typed git tag -a, but you didn't (and neither did lazygit, of course). It sounds like there might be a git config that makes -a the default for every tag, but I couldn't find any such option. Weird.

I went in search of the same kind of setting and didn't find one. I am hoping that someone here can suggest where to look. I went to the various git configuration files and saw nothing obvious. I'm searching all my dotfiles right now.

**Comment by maulik13** [comment]
I have the same issue, my final error prints a NANO editor UI but with jumbled text and an error. 

![image](https://github.com/user-attachments/assets/a36dd1f0-2a35-4be1-ba1e-332f466ac024)

Is there no way to create an annotated tag?

**Comment by stefanhaller** [comment]
> Is there no way to create an annotated tag?

Yes there is, as explained [above](https://github.com/jesseduffield/lazygit/issues/3423#issuecomment-2016833248).

I still have no idea why an editor is being opened for you. I can't explain this. I just tried it again, works just fine for me.

**Comment by stefanhaller** [comment]
@maulik13 Good find! That explains it, I can reproduce it when I set `tag.gpgSign` to true. (In my case I don't have a problem though, because my editor is a GUI program, so launching it at that point is not an issue; but I can see how it's a problem for people using an editor that runs in the terminal.)

@jbrains, can you confirm that you also have `tag.gpgSign` turned on?

As for fixing it, two things need to happen:
- when creating an annotated tag (i.e. the description is non-empty) we need to make it possible for people to type whatever gpg passphrase they might be asked for, if they are not using an agent. This involves replacing the `WithWaitingStatus` in `OpenCreateTagPrompt` with something like `gpgHelper.WithGpgHandling`, similar to how we do this in `handleCommit`. I say "something like" because it needs a bit of refactoring; `WithGpgHandling` currently only queries the `commit.gpgsign` config because it has only been used for committing so far, so we need to make this configurable somehow.
- when the description is empty, make it so that it doesn't try to open an editor for the message. There are two ways to do that:
  1. pass `--no-sign`
  2. pass `-m ""`
Both seem to work, the former creates a lightweight tag, the latter creates a signed annotated tag with an empty message. I tend to favor the first, since it's easier (we can simply pass the option unconditionally; for the other approach we have to first query whether signing is on), and also because the behaviour makes more sense to me. After all, the description field is our equivalent to opening the editor for a tag message, so it makes sense to require users to type something there if they want an annotated (and signed) tag.

Anybody interested in taking this on?

> I wish there was an option to do an annotated tag with an empty message from lazygit which executes `git tag -a -m v2.2.2 <hash>`

I'm not sure I understand why you want lazygit to make up this message for you. If `v2.2.2` is the tag message you want, why don't you type it in the description field?

**Comment by jbrains** [comment]
I can't confirm it for over a week, because I'm traveling laptop-free, but
I estimate the probability at above 99%. I know that I made GPG signing of
my commits the default some time in the past year or two. You can safely
assume that I am affected by the same cause

**Comment by stk-ableton** [comment]
> Can the tag creation detect `tag.gpgSign` is true and show the status so I know I have to provide a message? Or provide options to either leave message empty/copy-commit-hash, or turn off signing?

Yes, these are the options that I was listing in my second bullet point in [this comment above](https://github.com/jesseduffield/lazygit/issues/3423#issuecomment-2501788723). Somebody just needs to do the work. (I don't have time myself.)

**Comment by ChrisMcD1** [comment]
Duplicate of https://github.com/jesseduffield/lazygit/issues/2955

**Comment by ChrisMcD1** [comment]
/close

Resolved by https://github.com/jesseduffield/lazygit/pull/4394
